protocol/shannon/http_client.go

```diff
@@ -0,0 +1,297 @@
+	trace := createHTTPTrace(metrics)
+	tracedCtx := httptrace.WithClientTrace(ctx, trace)
+
+	// Return recorder function that logs request details.
+	requestRecorder := func(err error) {
+		atomic.AddInt64(&h.activeRequests, -1)
+		metrics.TotalTime = time.Since(metrics.StartTime)
+		metrics.Error = err
+		if err != nil {
+			h.logRequestMetrics(logger, *metrics)
+		}
+	}
+
+	return tracedCtx, requestRecorder
+}
+
+// categorizeError categorizes HTTP client errors and updates counters for monitoring
+func (h *httpClientWithTracing) categorizeError(ctx context.Context, err error) error {
+	if ctx.Err() == context.DeadlineExceeded {
+		atomic.AddInt64(&h.timeoutErrors, 1)
+		return fmt.Errorf("request timeout: %w", err)
+	} else {
+		atomic.AddInt64(&h.connectionErrors, 1)
+		return fmt.Errorf("connection error: %w", err)
+	}
+}
+
+// readAndValidateResponse reads the response body and validates the HTTP status code
+func (h *httpClientWithTracing) readAndValidateResponse(resp *http.Response) ([]byte, error) {
+	// Read response body with size protection
+	const maxResponseSize = 100 * 1024 * 1024 // 100MB limit
+	limitedReader := io.LimitReader(resp.Body, maxResponseSize)
+	responseBody, err := io.ReadAll(limitedReader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	// Validate HTTP status code
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
+		return nil, fmt.Errorf("%w: %d", errRelayEndpointHTTPError, resp.StatusCode)
+	}
+
+	return responseBody, nil
+}
+
+// createHTTPTrace creates an HTTP trace that captures timing metrics
+// for each phase of the HTTP request lifecycle.
+func createHTTPTrace(metrics *requestMetrics) *httptrace.ClientTrace {
+	var dnsStart, connectStart, tlsStart time.Time
+
+	return &httptrace.ClientTrace{
+		DNSStart: func(info httptrace.DNSStartInfo) {
+			dnsStart = time.Now()
+		},
+		DNSDone: func(info httptrace.DNSDoneInfo) {
+			if !dnsStart.IsZero() {
+				metrics.DNSLookupTime = time.Since(dnsStart)
+			}
+		},
+		ConnectStart: func(network, addr string) {
+			connectStart = time.Now()
+		},
+		ConnectDone: func(network, addr string, err error) {
+			if !connectStart.IsZero() {
+				metrics.ConnectTime = time.Since(connectStart)
+			}
+		},
+		TLSHandshakeStart: func() {
+			tlsStart = time.Now()
+		},
+		TLSHandshakeDone: func(state tls.ConnectionState, err error) {
+			if !tlsStart.IsZero() {
+				metrics.TLSTime = time.Since(tlsStart)
+			}
+		},
+		GotFirstResponseByte: func() {
+			metrics.FirstByteTime = time.Since(metrics.StartTime)
+		},
+	}
+}
+
+// logRequestMetrics logs comprehensive request metrics for debugging failed requests.
+// Only called when a request fails to avoid verbose logging on successful requests.
+func (h *httpClientWithTracing) logRequestMetrics(logger polylog.Logger, metrics requestMetrics) {
+	// Log detailed failure metrics using the provided structured logger
+	logger.With(
+		"url", metrics.URL,
```

Comment:
Thoughts on prefixing everything here with the same prefix?
- `http_client_debug_url`
- `http_client_with_tracing_url`
- `http_client_with_debug_metrics_url`